internal/infrastructure/kubernetes/ratelimit/resource.go

```diff
@@ -5,10 +5,117 @@
+	// LogLevelEnvVar is the log level.
+	LogLevelEnvVar = "LOG_LEVEL"
+	// UseStatsdEnvVar is the use statsd.
+	UseStatsdEnvVar = "USE_STATSD"
+	// InfraName is the name for rate-limit resources.
+	InfraName = "envoy-ratelimit"
+	// InfraGRPCPort is the grpc port that the rate limit service listens on.
+	InfraGRPCPort = 8081
+)
 
 // GetServiceURL returns the URL for the rate limit service.
 // TODO: support custom trust domain
 func GetServiceURL(namespace string) string {
 	return fmt.Sprintf("grpc://%s.%s.svc.cluster.local:%d", InfraName, namespace, InfraGRPCPort)
 }
+
+// rateLimitLabels returns the labels used for all envoy rate limit resources.
+func rateLimitLabels() map[string]string {
+	return map[string]string{
+		"app.gateway.envoyproxy.io/name": InfraName,
+	}
+}
+
+// expectedRateLimitContainers returns expected rateLimit containers.
+func expectedRateLimitContainers(ratelimit *egcfgv1a1.RateLimit, rateLimitDeployment *egcfgv1a1.KubernetesDeploymentSpec) []corev1.Container {
+	ports := []corev1.ContainerPort{
+		{
+			Name:          "http",
+			ContainerPort: InfraGRPCPort,
+			Protocol:      corev1.ProtocolTCP,
+		},
+	}
+
+	containers := []corev1.Container{
+		{
+			Name:            InfraName,
+			Image:           *rateLimitDeployment.Container.Image,
+			ImagePullPolicy: corev1.PullIfNotPresent,
+			Command: []string{
+				"/bin/ratelimit",
+			},
+			Env:             expectedRateLimitContainerEnv(ratelimit, rateLimitDeployment),
+			Ports:           ports,
+			Resources:       *rateLimitDeployment.Container.Resources,
+			SecurityContext: rateLimitDeployment.Container.SecurityContext,
+			VolumeMounts: []corev1.VolumeMount{
+				{
+					Name:      InfraName,
+					MountPath: "/data/ratelimit/config",
+					ReadOnly:  true,
+				},
+			},
+			TerminationMessagePolicy: corev1.TerminationMessageReadFile,
+			TerminationMessagePath:   "/dev/termination-log",
+		},
+	}
+
+	return containers
+}
+
+// expectedRateLimitContainerEnv returns expected ratelimit container envs.
+func expectedRateLimitContainerEnv(ratelimit *egcfgv1a1.RateLimit, rateLimitDeployment *egcfgv1a1.KubernetesDeploymentSpec) []corev1.EnvVar {
+	env := []corev1.EnvVar{}
+
+	defaultEnvMapping := map[string]string{
+		RedisSocketTypeEnvVar:       "tcp",
+		RuntimeRootEnvVar:           "/data",
+		RuntimeSubdirectoryEnvVar:   "ratelimit",
+		RuntimeIgnoreDotfilesEnvVar: "true",
+		RuntimeWatchRootEnvVar:      "false",
+		LogLevelEnvVar:              "info",
+		UseStatsdEnvVar:             "false",
+	}
+
+	for _, envVar := range rateLimitDeployment.Container.Env {
+		// extension env provides current env, remove from default env mapping
+		if defaultEnvMapping[envVar.Name] != "" {
+			delete(defaultEnvMapping, envVar.Name)
+		} else {
+			env = append(env, envVar)
+		}
+	}
+
+	// append default env that extension env does not provide.
+	for key, value := range defaultEnvMapping {
+		env = append(env, corev1.EnvVar{Name: key, Value: value})
```

Comment:
@arkodg there use the map so sort env, to make sure the test data is.right

Reply:
there is a 🐛, i will fix it today

Reply:
cool, will review again after